### PR TITLE
chore(recordings): Optimize recordings list query

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem
- addressing https://github.com/PostHog/posthog/issues/14072

## Changes

- implements suggestions from https://github.com/PostHog/posthog/issues/14072

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Comment by EDsCODE** [comment]
TODO:
- [ ] handle person properties pushdown

File: posthog/queries/session_recordings/session_recording_list.py [hunk 1]
```diff
@@ -406,3 +435,143 @@ def run(self, *args, **kwargs) -> SessionRecordingQueryResult:
...
+            non_aggregate_select_condition_clause,
+            aggregate_event_select_clause,
+            aggregate_select_clause,
+            aggregate_having_clause,
+            where_conditions,
+            params,
+        )
+
+    def run(self, *args, **kwargs) -> SessionRecordingQueryResult:
+        self._filter.hogql_context.using_person_on_events = True
```
Comment (EDsCODE) [inline posthog/queries/session_recordings/session_recording_list.py:573]: This connection to hogql restricts the improvement to person_on_events enabled orgs for now.

**Comment by EDsCODE** [comment]
pretty sure remaining test failures are unrelated

File: posthog/queries/session_recordings/session_recording_list.py (hunk 1 above)
Comment (neilkakkar) [reply posthog/queries/session_recordings/session_recording_list.py:573]: don't get this, aren't there like only 4 poe orgs right now?

File: posthog/queries/session_recordings/test/__snapshots__/test_session_recording_list.ambr [hunk 2]
```diff
@@ -1172,3 +1052,1109 @@
...
+            any(distinct_id) as distinct_id,
+            SUM(has_full_snapshot) as full_snapshots
+     FROM session_recording_events PREWHERE team_id = 2
+     AND timestamp >= '2021-01-13 12:00:00'
+     AND timestamp <= '2021-01-22 08:00:00'
+     GROUP BY session_id
+     HAVING full_snapshots > 0
+     AND start_time >= '2021-01-14 00:00:00'
+     AND start_time <= '2021-01-21 20:00:00') AS session_recordings ON session_recordings.session_id = events.session_id
+  WHERE session_recordings.distinct_id == events.distinct_id
```
Comment (neilkakkar) [inline posthog/queries/session_recordings/test/__snapshots__/test_session_recording_list.ambr:1113]: doesn't matching by distinctID lead to problems as we're randomly selecting one?

Don't think sessionIDs change between login steps, but distinctIDs can & do?

**Review by neilkakkar — APPROVED** [review]
looks good! The code here is going get more complicated as we support multiple ways of doing things, but given that it will all go away _sometime_ with the schema changes, not too bothered.

Would like it if session-recording folks have a look too though!

File: posthog/models/team/team.py [hunk 3]
```diff
@@ -217,6 +217,26 @@ def _person_on_events_querying_enabled(self) -> bool:
         # on self-hosted, use the instance setting
         return get_instance_setting("PERSON_ON_EVENTS_ENABLED")
 
+    @property
+    def recordings_list_v2_query_enabled(self) -> bool:
```
Comment (benjackwhite) [inline posthog/models/team/team.py:221]: Could we do this on the frontend instead and control it via a query param to the API? Makes it much easier to test back and forth in a live situation to compare results etc.

File: posthog/queries/session_recordings/test/__snapshots__/test_session_recording_list.ambr (hunk 2 above)
Comment (benjackwhite) [reply posthog/queries/session_recordings/test/__snapshots__/test_session_recording_list.ambr:1113]: I also wonder this... Seems like a copy from the v1 query but I'm not sure why it matters... We only need one distinctId to load the Person anyways and we have that from the recording. We anyways group by session_id... 

Perhaps this is to account for weird cases where there is a clash of session_id but I don't see that happening...

File: posthog/queries/session_recordings/session_recording_list.py
Comment (benjackwhite) [inline posthog/queries/session_recordings/session_recording_list.py:1]: Overall I think this is worth trying but reading it still makes me think "there must be a simpler way" :D 

I keep thinking why is not just 
```
select * from recording_events_grouped where session_id in (BuildStandardEventsQuery(filters).select("session_id")
```

Definitely don't want to derail and at this point I'll take any improvements to the speed of this thing but these queries still confuse the hell out of me

**Review by benjackwhite — CHANGES_REQUESTED** [review]

File: posthog/queries/session_recordings/session_recording_list.py (hunk 1 above)
Comment (EDsCODE) [reply posthog/queries/session_recordings/session_recording_list.py:573]: yeah. hogql allows for person properties to be filtered in the recordings event query which usually was not the case. Now, either we use person on events and only have this optimization working for the tiny subset until everything works, or I go through a refactor that splits up and does the right pushdown of person property filtering to the subquery when we detect a hogql based person property 
Screenshot 2023-03-14 at 9 52 03 AM

Comment (neilkakkar) [reply posthog/queries/session_recordings/session_recording_list.py:573]: Ugh I see, okay, good enough

File: posthog/models/team/team.py (hunk 3 above)
Comment (EDsCODE) [reply posthog/models/team/team.py:221]: Hmm, on second thought, I don't want to add filters to the object and muddy our caching for insights. You think it's worth it for this?

File: posthog/queries/session_recordings/session_recording_list.py
Comment (mariusandra) [reply posthog/queries/session_recordings/session_recording_list.py:1]: It should be that easy... why isn't it? :D

What would this look like if you rebuilt it the same way [the events query](https://github.com/PostHog/posthog/blob/3d3a5666323b25cdaad917806c8f7460ef2ab80a/posthog/models/event/events_query.py) is built?

Comment (EDsCODE) [reply posthog/queries/session_recordings/session_recording_list.py:1]: Bigger issue to wait on before spending time unnecessarily is person on events deployment. Once we're there, we can probably move this to @mariusandra's suggestion with less headache and a lot less joins to negotiate

**Comment by EDsCODE** [comment]
Will need this to wait on person on events because of this dependency: https://github.com/PostHog/posthog/pull/14458#discussion_r1122476083

**Comment by mariusandra** [comment]
The point of HogQL is that it'll deal with the persons on events issue for you. I'm [actively working on](https://github.com/PostHog/posthog/pull/14740) making this actually true. Once that PR is wrapped up (tomorrow? 🤞), person joins will and won't happen as needed, for both the `events` and `session_recording_events` tables. You can just use `person_id` and `person.properties.foo` as if it's on the same table.

I also wrote up a [quick guide](https://github.com/PostHog/posthog.com/pull/5520/files?short_path=1af2b47#diff-1af2b47178721624553549d092a7b7bfcc9dff5153460e614936a31fcab21835) on how to use HogQL with Python in the backend. Hopefully that can provide some inspiration. You will be free to do any joins or non-joins as you'd like, or let HogQL handle it for you. It's just a layer around ClickHouse SQL after all... as long as you stick to the [implemented syntax](https://github.com/PostHog/meta/issues/81) (no arrays yet :/).

**Comment by EDsCODE** [comment]
I've updated the logic to not use V2 if there are hogql properties. This will be in effect until person on events is fully deployed. The updates to hogql mentioned do not solve this particular issue because the recordings query does adhoc property parsing that does not adhere to the rest of the query patterns

File: posthog/models/team/team.py (hunk 3 above)
Comment (benjackwhite) [reply posthog/models/team/team.py]: I wasn't suggesting adding it to filters (not everything has to go in that mega object 😅 ) - just a standard query param. I pushed the change in the way I meant so now we just check a query param which makes it super easy to flip back and forth without having to change the feature flag all the time

**Review by benjackwhite — APPROVED** [review]
Overall I'm happy for this to go in but it would be great if this doesn't hang around for 3 months as half-done thing 😅 Ideally we test it heavily and immediately, and then follow up with a review to swap over fully ASAP. 

In the not too far future, we're going to be rebuilding the whole thing (hopefully using the new query stuff) in order to support Notebooks more effectively so it would be good to have as a clean a base to start from

**Pull request merged**